src/Coordination/KeeperStateMachine.cpp

```diff
@@ -523,7 +523,7 @@ void KeeperStateMachine::processReadRequest(const KeeperStorage::RequestForSessi
         true /*check_acl*/,
         true /*is_local*/);
     for (const auto & response : responses)
-        if (!responses_queue.push(response))
+        if (!responses_queue.push(response) && !responses_queue.isFinished())
```

Comment:
Maybe it is better to move this check out from loop?

Reply:
To address this and the comment above, the biggest problem is the NuRaft.
If I throw an exception from one of its threads, `std::abort` happens (that's what caused the problem), that lib doesn't handle exceptions nicely as CH. Assume all methods here are called by NuRaft and not CH.

The queue can be finished in between the loop iterations but it can also be finished between the checks:
```
if finished -> finished is false
  break;
if (!push(...)) -> finished is true, i.e we have same problem
```
I can add the check for break to not do unnecessary iterations as @novikd suggested, but because I would still need to keep the check after an unsuccessful push I wouldn't change much code-wise.

So I would like to merge this to fix CI ASAP and then polish it.